Create spec.md (#15)
deltap2p/hobbits: specs/spec.md

# @TODO
- [ ] Link to parsers.
- [ ] attestations

# Hobbits

*Written by Dean Eigenmann <dean@status.im> & Rene Nayman <rene@whiteblock.io>* 

## Table of Contents

1. [Abstract](#abstract)
2. [Flow](#flow)
3. [Wire Protocol](#wire-protocol)
    3.1 [Messages](#messages)
4. [Transport](#transport)
5. [Protocols](#protocols)
    5.1 [RPC](#rpc)
    5.2 [Gossip](#gossip)
    5.3 [PING](#ping)
6. [Implementations](#implementations)

## Abstract

Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.

Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.

## Flow

We layout a network where the nodes are peered in the following manner:

![nodes](./nodes.png)

The following flow dictates how blocks are shared amongst peers:

![sequence diagram](./diagram.png)

> 1. Node 1 receives a new block `OxCAB` and gossips it to Node 2 and Node 3

Reviewer: ```suggestion
1. Node 1 receives a new block `0xCAB` and gossips `hash(0xCAB)` to Node 2 and Node 3
```

Can we make it clear that we are gossiping hashes and not the full block? In this workflow.

> 2. Node 2 requests the block bodies for block `OxCAB` from Node 1 in an RPC Command query

Reviewer: I think you wrote `Ox` instead of `0x`. OH vs ZERO. The former is not a valid hex string.

Here and everywhere.

3. Node 1 responds to Node 2 with the block bodies for block `OxCAB` in an RPC Command response
4. Node 2 receives a new block `OxCAB` and gossips it to Node 1 and Node 3

## Wire Protocol

### Messages

> Messages are sent as `utf-8 byte encoded arrays`. The `message` format looks as follows:

Reviewer: `utf-8 byte encoded arrays`? Is this a fancy way to write "strings"? 

Anything you can do here to reduce ambiguity would be nice.

```
EWP <version> <protocol> <header-length> <body-length>
<header><body>
```

**NOTE:** Message`body`s along with `header` fields are BSON encoded.

#### Fields

Every hobbit message contains the following fields: 

| Field | Definition | Validity |
|:------:|----------|:----:|
| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |

> | `protocol` | Defines the [protocol](#protocol). | `(RPC\|GOSSIP\|PING)` |

Reviewer: Are these enum values represented as int or literately "RPC" as string.

Reviewer: This link is broken.

Reviewer 2: I would assume as numbers...

Author: nah, strings.

> | `protocol` | Defines the [protocol](#protocol). | `(RPC\|GOSSIP\|PING)` |

Reviewer 2: ```suggestion
| `protocol` | Defines the [protocol](#protocols). | `(RPC\|GOSSIP\|PING)` |
```

| `header` | Defines the header | payload |
| `body` | Defines the body | payload |

## Transport

<!-- @TODO DISCUSS TCP AND SHIT -->

## Protocols

Hobbits defines 3 protocols that dictate how messages are interpreted and responded to.

### RPC

<!-- @TODO WRITE SOME SHIT -->

> #### Envelope

Reviewer: Is it helpful to have the ID of the message we are responding to? 

I can't think of a use case yet, but I thought that might be something to add if you are asking the peer for several similar request then we might want to know which one they are responding to.

Reviewer 3: That's in the envelope below `requestID`

```python
{ 

>   'method_id': 'uint16' ## byte representing the method

Reviewer: This is an enum value (backed by uint16). Not sure if it helps to clarify that here.

Also, you wrote "byte representing the method" but this is misleading because 16 bit integer is actually 2 bytes.

Reviewer 3: Taken from Matt Slipper's thinking: less bytes, better. We can explain in more detail, I'll open an issue.

>   'id': 'uint64' ## id of the request

Reviewer: Want to make this a string? That would support uuid or somehow baking the peer ID into the ID.

Reviewer 3: Fine with a string. Again, there was a thinking around limiting number of bytes there, which is probably not a concern.

Reviewer 3: Opened  an issue.

  'body': 'bytes' ## body of the request itself
}
```

The `body` field contains an [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md) encoded byte array. 

##### Example

Below you will find an example `RPC` request.

```
EWP 0.2 RPC 0 65
{
  "method_id": 0x01,
  "id": 1,
  "body": {
    "bodies": "0x2a00000000000000"
  }
}
```
#### RPC Methods

There are two types of RPC methods: one that requests data and one that responds with data.

* Peers may request blocks and headers from other peers, possibly in response to a `GOSSIP` message. 
* Peers may request blocks repeatedly from the same peers.
* Other peers may respond on a best effort basis with header and block data.
* There is no SLA for responding. 

##### Handshake

###### `0x00` HELLO

> Upon discovering each other, nodes may exchange `HELLO` messages.

Reviewer: ```suggestion
Upon discovering each other, nodes should exchange `HELLO` messages.
```

https://tools.ietf.org/html/rfc2119

If you want to be extra fancy, mention that this doc confirms to the best practices terminology in RFC2119 at the top of the doc.

Reviewer 3: The intent here was to show the optionality of the hello message exchange. In the case of RLPx, a HELLO message must be exchanged. We are ok with hobbits having peers jump directly to asking headers for example.

> Nodes may send `HELLO` to other peers when they exchange messages for the first time or when their state changes to let them know new blocks are available.

Reviewer 4: Don't we already let peers know that new blocks are available with GOSSIP anyway?

Reviewer 4: Oh is this to let them know about the new (fork choice) head? rather than the new gossipped block?

Reviewer 3: It's when you just come up the first time. You don't know anything about the state of the network and you need to sync up.

Nodes can send `HELLO` messages to each other to exchange information on their status:

```python
{
  'network_id': 'uint8' ## the ID of the network (1 for mainnet, and some predefined number for a testnet)
  'chain_id': 'uint8' ## the ID of the chain (1 for ETH)
  'latest_finalized_root': 'bytes32' ## the hash of the latest finalized root
  'latest_finalized_epoch': 'uint64' ## the number of the latest finalized epoch
  'best_root': 'bytes32' ## the hash of the best root this node can offer
  'best_slot': 'uint64' ## the number of the best slot this node can offer
}
```

> Upon receiving a `HELLO` message, the node may reply with a `HELLO` message.

Reviewer: ```suggestion
Upon receiving a `HELLO` message, the node should reply with a `HELLO` message.
```

How else are they going to get this data?

###### `0x01` GOODBYE

Nodes may signal to other nodes that they are going away by sending a `GOODBYE` message:

```python
{
  'reason': 'uint64' ## an optional reason code up to the client
}
```

The reason given is optional. Reason codes are up to each client and should not be trusted.

Upon receiving a `GOODBYE` message, no response is necessary.

##### `0x02` GET_STATUS

Nodes may exchange metadata information using a `GET_STATUS` message.
<!--is this necessary?-->

> A `GET_STATUS` request may be sent in response to receiving a `GOSSIP` message:

Reviewer 4: This actually might be a good place to return (fork choice) head block.

Author: @jrhea

Reviewer 3: @cemozerr open a new issue and add more fields to status. You can throw a bunch of metadata in there.

  'hash_signature': 'bytes32' ## a signature of the message hash with a public key identifying the node sending data
}
```

##### Example
```
EWP 0.2 GOSSIP 222 0
{
  "method_id": 3,
  "topic": "BLOCK",
  "timestamp": 1560471980,
  "message_hash": "0x9D686F6262697473206172652074776F20616E6420666F75722066656574",
  "hash_signature": "0x0000000009A4672656E63682070656F706C6520617265207468652062657374"
}
```

The message may contain additional headers specified by the application layer.

#### GOSSIP Methods

<!--should we include examples of messages?-->
<!--should we include that RPC commands can be a response to IHAVE message?-->

##### `0x00` GOSSIP

Nodes use `GOSSIP` methods to send data to other nodes in the network.

The body of a `GOSSIP` method consists in the data being gossiped and 

**@TODO: is the body an ssz encoded byte array?**

The `message_hash` header value must match the hash of the contents of the body according to a predefined hash function defined by the application.

### Ping

The ping/pong protocol is used to test connections between two peers and ensure the Hobbits implementation is passing conformance tests.

> When a `PING` message is received, the node should respond with the body of the `PING` message as a `PONG` message.

Reviewer: ```suggestion
When a `PING` message is received, the node must respond with the body of the `PING` message as a `PONG` message.
```

# @TODO
- [ ] Link to parsers.
- [ ] attestations
- [ ] what else needs to be defined in the transport header?

# Hobbits

> *Written by Dean Eigenmann <dean@status.im> & Rene Nayman <rene@whiteblock.io>* 

Reviewer 3: We can move acknowledgements to the very end of the doc. Let's make sure Matt Elder gets in there too.

## Table of Contents

1. [Abstract](#abstract)
2. [Flow](#flow)
3. [Wire Protocol](#wire-protocol)
    - [Messages](#messages)
4. [Transport](#transport)
5. [Protocols](#protocols)
    - [RPC](#rpc)
    - [Gossip](#gossip)
    - [PING](#ping)
6. [Implementations](#implementations)

## Abstract

Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.

Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.

## Flow

We layout a network where the nodes are peered in the following manner:

![nodes](./nodes.png)

The following flow dictates how blocks are shared amongst peers:

![sequence diagram](./diagram.png)

> 1. Node 1 receives a new block `0xCAB` and gossips `hash(0xCAB)` to Node 2 and Node 3

Reviewer 5: it might be worth it to define what this hashing algorithm is `hash(0xCAB)`

Author: @atoulme do we know what the hashing algo is yet? Otherwise I'll just pick one...

Reviewer 3: We picked sha256 iirc (or maybe keccak). Pick something you like.

2. Node 2 requests the block bodies for block `0xCAB` from Node 1 in an RPC Command query
3. Node 1 responds to Node 2 with the block bodies for block `0xCAB` in an RPC Command response
4. Node 2 receives a new block `0xCAB` and gossips it to Node 1 and Node 3

## Wire Protocol

### Messages

The `message` format looks as follows:

```
EWP <version> <protocol> <header-length> <body-length>
<header><body>
```

> **NOTE:** Message`body`s along with `header` fields are BSON encoded.

Reviewer 6: Is this true?  I mean, required?  The wire protocol just says, arbitrary binary data, for the header and body; and there's header-length and body-length to tell how to extract it.  

Maybe the BSON-encoding might happen in a layer above the wire-protocol, in GOSSIP/etc?  I'm not sure that it's required here.  

Similar comment about the "messages are utf-8 encoded strings" line, too... that would rule out binary payloads unless they're encoded first

Reviewer 3: That’s how it was before. But we’re trying to simplify and cram all the details into a minimum doc so we can get a bare minimum interop going.

#### Fields

Every hobbit message contains the following fields: 

| Field | Definition | Validity |
|:------:|----------|:----:|
| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |
| `protocol` | Defines the [protocol](#protocols). | `(RPC\|GOSSIP\|PING)` |
| `header` | Defines the header | payload |
| `body` | Defines the body | payload |

## Transport

A Hobbits node listens for and sends out messages. 

Hobbits packets are sent as follows: 

```
<length><packet>
```

- **Length**: `big endian` encoded `uint32` that represents the length of the packet, with a length of 4 bytes.
- **Packet**: Byte encoded string of the hobbits message.

## Protocols

Hobbits defines 3 protocols that dictate how messages are interpreted and responded to.

### RPC

<!-- @TODO WRITE SOME SHIT -->

#### Envelope

```python
{ 
  'method_id': 'uint16' ## byte representing the method
  'id': 'uint64' ## id of the request
  'body': 'bytes' ## body of the request itself
}
```

The `body` field contains an [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md) encoded byte array. 

##### Example

> Below you will find an example `RPC` request.

Reviewer 5: isn't this a response? I thought `bodies` contains an ssz encoded byte away

Reviewer 3: OK should be message instead.

Reviewer 3: ```suggestion
Below you will find an example `RPC` message.
```

```
EWP 0.2 RPC 0 65
{
  "method_id": 0x01,
  "id": 1,
  "body": {
    "bodies": "0x2a00000000000000"
  }
}
```
#### RPC Methods

There are two types of RPC methods: one that requests data and one that responds with data.

* Peers may request blocks and headers from other peers, possibly in response to a `GOSSIP` message. 
* Peers may request blocks repeatedly from the same peers.
* Other peers may respond on a best effort basis with header and block data.
* There is no SLA for responding. 

##### Handshake

###### `0x00` HELLO

Upon discovering each other, nodes may exchange `HELLO` messages.

Nodes may send `HELLO` to other peers when they exchange messages for the first time or when their state changes to let them know new blocks are available.

Nodes can send `HELLO` messages to each other to exchange information on their status:

```python
{
  'network_id': 'uint8' ## the ID of the network (1 for mainnet, and some predefined number for a testnet)
  'chain_id': 'uint8' ## the ID of the chain (1 for ETH)
  'latest_finalized_root': 'bytes32' ## the hash of the latest finalized root
  'latest_finalized_epoch': 'uint64' ## the number of the latest finalized epoch

>   'best_root': 'bytes32' ## the hash of the best root this node can offer

Reviewer 5: ```suggestion
  'best_root': 'bytes32' ## the hash of the best block root this node can offer
```

Reviewer: Why did you undo this suggestion?

}
```

Upon receiving a `HELLO` message, the node should reply with a `HELLO` message.

###### `0x01` GOODBYE

Nodes may signal to other nodes that they are going away by sending a `GOODBYE` message:

```python
{
  'reason': 'uint64' ## an optional reason code up to the client
}
```

The reason given is optional. Reason codes are up to each client and should not be trusted.

Upon receiving a `GOODBYE` message, no response is necessary.

##### `0x02` GET_STATUS

Nodes may exchange metadata information using a `GET_STATUS` message.
<!--is this necessary?-->
A `GET_STATUS` request may be sent in response to receiving a `GOSSIP` message:

```python
{
  'user_agent': 'bytes' ## the human readable name of the client, optionally with its version and other metadata
  'timestamp': 'uint64' ## the current time of the node in milliseconds since epoch
}
```

Any peer may provide information about their status and metadata to any other peer. Other peers may respond on a best effort basis, if at all.

##### Block Headers

###### `0x0A` GET_BLOCK_HEADERS

Nodes may request block headers from other nodes using the `GET_BLOCK_HEADERS` message:

> <!--is this necessary?-->

Reviewer 5: I don't think this is necessary yet

```python
{
  'start_root': 'bytes32' ## the root hash to start querying from OR
  'start_slot': 'uint64' ## the slot number to start querying from
  'max': 'uint64' ## the max number of elements to return
  'skip': 'uint64' ## the number of elements apart to pick from
  'direction': 'uint8' ## 0x01 is ascending, 0x00 is descending direction to query elements
}
```

A `GET_BLOCK_HEADERS` request may be sent in response to receiving a `GOSSIP` message.

###### `0x0B` BLOCK_HEADERS

> Nodes may provide block roots to other nodes using the `BLOCK_HEADERS` message, usually in response to a `GET_BLOCK_HEADERS` message:

Reviewer 5: Why are the headers necessary as network objects? 
I think `GET_BLOCK_BODIES` and `BLOCK_BODIES` are sufficient enough. `BeaconBlock` contains header

Reviewer 3: Same as LES. You sync headers fast, and bodies separately at a slower pace. Having them separated allows a number of other use cases such as light clients.

1. [Abstract](#abstract)
2. [Flow](#flow)
3. [Wire Protocol](#wire-protocol)
    1. [Messages](#messages)
4. [Transport](#transport)
5. [Protocols](#protocols)

>     1. [RPC](#rpc)

Reviewer: protip: if you make all of these `1` then github will automatically number them for you in rendering. That way you don't need to go change all of the numbers if you add/remove in the middle of the list or reorder things.

Author: I use an editor that auto updates the numbers, so its like that when I copy over. Thanks for the tip tho @prestonvanloon
